include/material/modified_cam_clay.tcc

```diff
@@ -0,0 +1,722 @@
+  // Compute pcd
+  (*state_vars).at("pcd") = mc_a_ * pow((*state_vars).at("chi") * s_h_, mc_b_);
+  // Compute pcc
+  (*state_vars).at("pcc") = mc_c_ * pow((*state_vars).at("chi") * s_h_, mc_d_);
+}
+
+//! Compute subloading parameters
+template <unsigned Tdim>
+void mpm::ModifiedCamClay<Tdim>::compute_subloading_parameters(
+    const double subloading_r, mpm::dense_map* state_vars) {
+  const double p = (*state_vars).at("p");
+  // Preconsolidation pressure
+  const double pc = (*state_vars).at("pc");
+  // Get bonding parameters
+  const double pcd = (*state_vars).at("pcd");
+  const double pcc = (*state_vars).at("pcc");
+  // Plastic strain
+  const double dpvstrain = (*state_vars).at("dpvstrain");
+  const double dpdstrain = (*state_vars).at("dpdstrain");
+  // Initialise subloading surface ratio
+  if ((*state_vars).at("subloading_r") == 1.0)
+    (*state_vars).at("subloading_r") = p / (pc + pcd + pcc);
+  else
+    // Update subloading surface ratio
+    (*state_vars).at("subloading_r") =
+        subloading_r -
+        subloading_u_ * (1 + (pcd + pcc) / pc) * log(subloading_r) *
+            std::sqrt(dpvstrain * dpvstrain + dpdstrain * dpdstrain);
+  // Threshhold
+  if ((*state_vars).at("subloading_r") < std::numeric_limits<double>::epsilon())
+    (*state_vars).at("subloading_r") = 1.E-5;
+  if ((*state_vars).at("subloading_r") > 1.)
+    (*state_vars).at("subloading_r") = 1.;
+}
+
+//! Compute dF/dmul
+template <unsigned Tdim>
+void mpm::ModifiedCamClay<Tdim>::compute_df_dmul(
+    const mpm::dense_map* state_vars, double* df_dmul) {
+  // Stress invariants
+  const double p = (*state_vars).at("p");
+  const double q = (*state_vars).at("q");
+  const double m_theta = (*state_vars).at("m_theta");
+  // Preconsolidation pressure
+  const double pc = (*state_vars).at("pc");
+  // Get bonding parameters
+  const double pcd = (*state_vars).at("pcd");
+  const double pcc = (*state_vars).at("pcc");
+  // Get elastic modulus
+  const double e_b = (*state_vars).at("bulk_modulus");
+  const double e_s = (*state_vars).at("shear_modulus");
+  // Get consistency parameter
+  const double mul = (*state_vars).at("delta_phi");
+  // Compute dF / dp
+  double df_dp = 2 * p - pc - pcd;
+  // Compute dF / dq
+  double df_dq = 2 * q / pow(m_theta, 2);
+  // Compute dF / dpc
+  double df_dpc = -(p + pcc);
+  // Upsilon
+  double upsilon = (1 + (*state_vars).at("void_ratio")) / (lambda_ - kappa_);
+  // A_den
+  double a_den = 1 + (2 * e_b + upsilon * (pc + pcd)) * mul;
+  // Compute dp / dmul
+  double dp_dmul = -e_b * (2 * p - pc - pcd) / a_den;
+  // Compute dpc / dmul
+  double dpc_dmul = upsilon * (pc + pcd) * (2 * p - pc - pcd) / a_den;
+  // Compute dq / dmul
+  double dq_dmul = -q / (mul + pow(m_theta, 2) / (6 * e_s));
+  // Compute dF / dmul
+  if (!bonding_)
+    (*df_dmul) = (df_dp * dp_dmul) + (df_dq * dq_dmul) + (df_dpc * dpc_dmul);
+  // Compute bonding parameters
+  if (bonding_) {
```

Comment:
you do two testings here. It's a better practice to if-else conditions

Reply:
Fix it